Framework/Algorithms/src/RebinToWorkspace.cpp

```diff
@@ -43,29 +45,97 @@ void RebinToWorkspace::init() {
+
+  // look for first non-equal x-axis between the workspaces
+  const size_t numHist = left->getNumberHistograms();
+  for (size_t i = 0; i < numHist; ++i) {
+    if (left->getSpectrum(i).readX() != rght->getSpectrum(i).readX()) {
+      return true;
+    }
+  }
+  // everything must be the same
+  return false;
+}
+} // namespace
+
 /**
  * Execute the algorithm
  */
 void RebinToWorkspace::exec() {
   // The input workspaces ...
   MatrixWorkspace_sptr toRebin = getProperty("WorkspaceToRebin");
   MatrixWorkspace_sptr toMatch = getProperty("WorkspaceToMatch");
-  bool PreserveEvents = getProperty("PreserveEvents");
-
-  // First we need to create the parameter vector from the workspace with which
-  // we are matching
-  std::vector<double> rb_params = createRebinParameters(toMatch);
-
-  IAlgorithm_sptr runRebin = createChildAlgorithm("Rebin");
-  runRebin->setProperty<MatrixWorkspace_sptr>("InputWorkspace", toRebin);
-  runRebin->setPropertyValue("OutputWorkspace", "rebin_out");
-  runRebin->setProperty("params", rb_params);
-  runRebin->setProperty("PreserveEvents", PreserveEvents);
-  runRebin->setProperty("IgnoreBinErrors", true);
-  runRebin->executeAsChildAlg();
-  progress(1);
-  MatrixWorkspace_sptr ws = runRebin->getProperty("OutputWorkspace");
-  setProperty("OutputWorkspace", ws);
+  bool preserveEvents = getProperty("PreserveEvents");
+
+  if (needToRebin(toRebin, toMatch)) {
+    // TODO should copy code from Rebin to do the right thing spectrum by
+    // spectrum
+    if (!WorkspaceHelpers::commonBoundaries(*toMatch)) {
+      throw std::runtime_error(
+          "WorkspaceToMatch must have common bin boundaries in all spectra");
+    }
+
+    // First we need to create the parameter vector from the workspace with
+    // which we are matching
+    std::vector<double> rb_params = createRebinParameters(toMatch);
+
+    IAlgorithm_sptr runRebin = createChildAlgorithm("Rebin");
+    runRebin->setProperty<MatrixWorkspace_sptr>("InputWorkspace", toRebin);
+    runRebin->setPropertyValue("OutputWorkspace", "rebin_out");
+    runRebin->setProperty("params", rb_params);
+    runRebin->setProperty("PreserveEvents", preserveEvents);
+    runRebin->setProperty("IgnoreBinErrors", true);
+    runRebin->setChildStartProgress(0.);
+    runRebin->setChildStartProgress(1.);
+    runRebin->executeAsChildAlg();
+    MatrixWorkspace_sptr outputWS = runRebin->getProperty("OutputWorkspace");
+    setProperty("OutputWorkspace", outputWS);
+  } else { // don't need to rebin
+    g_log.information("Workspaces have matched binning");
+    MatrixWorkspace_sptr outputWS = getProperty("OutputWorkspace");
+
+    bool inPlace = (toRebin == outputWS);
+    bool isEvents =
+        bool(boost::dynamic_pointer_cast<const EventWorkspace>(toRebin));
+
+    if (inPlace) {
+      if (isEvents && (!preserveEvents)) {
+        // convert to a MatrixWorkspace
+        IAlgorithm_sptr convert =
+            createChildAlgorithm("ConvertToMatrixWorkspace");
```

Comment:
Why do you need special handling for `EventWorkspace`? Rebinning as above should give the same result as `ConvertToMatrixWorkspace`?